### PR TITLE
NCVoter/ChangeTracker refactor and processor fix + tests!

### PR DESCRIPTION
**Pull request by ironfroggy**

### CODE REVIEW

File: scripts/mutate_dataset.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
+#!/usr/bin/env python3
```
Comment (vkurup) [inline scripts/mutate_dataset.py]: Was this file just for your testing? Or is it meant to be committed to the repo? If so, maybe we could add a bit of documentation on how a dev should use it?

File: setup.cfg [hunk 2]
```diff
@@ -1,3 +1,7 @@
 [flake8]
 max-line-length=160
 exclude=migrations,docs,downloads
+
+# H233: Python 3.x incompatible use of print operator
```
Comment (vkurup) [inline setup.cfg]: I think we can remove H233 (It's not triggered anywhere when I run flake8). I'd prefer to remove E251 to keep consistent with all of our other projects, but up to you if you feel strongly.

File: voter/management/commands/voter_process_snapshot.py [hunk 3]
```diff
@@ -48,23 +45,38 @@ def find_md5(row_data, exclude=[]):
...
 def get_file_lines(filename):
-    # ftfy lets us iterate over lines while it corrects encoding problems
-    lines = ftfy.fix_file(filename)
-
+    # guess the number of lines
+    f = open(filename, 'rb')
+    chunk = f.read(1024 * 1024)
+    newlines_per_meg = chunk.count(b'\n')
+    file_megs = os.stat(filename).st_size / (1024 * 1024)
+    line_count = file_megs * newlines_per_meg
```
Comment (vkurup) [inline voter/management/commands/voter_process_snapshot.py]: Could we call this `approx_line_count`?

File: voter/management/commands/voter_process_snapshot.py [hunk 4]
```diff
@@ -48,23 +45,38 @@ def find_md5(row_data, exclude=[]):
...
+
+    # UTF16 or (presumably) BOM-less UTF8
+    f.seek(0)
+    if f.read(2) == b'\xff\xfe':
+        encoding = 'utf16'
+    else:
+        encoding = 'latin1'
+
+    f = open(filename, encoding=encoding)
```
Comment (vkurup) [inline voter/management/commands/voter_process_snapshot.py]: Should we close the file before reopening? (or put the block above in a `with` context?)

File: voter/management/commands/voter_process_snapshot.py [hunk 5]
```diff
@@ -122,129 +126,139 @@ def reset_file(file_tracker):
     file_tracker.save()
 
 
-@transaction.atomic
+# @transaction.atomic
```
Comment (vkurup) [inline voter/management/commands/voter_process_snapshot.py]: Are you thinking we'll bring this back? Or should we just remove it?

File: voter/management/commands/voter_process_snapshot.py [hunk 6]
```diff
@@ -122,129 +126,139 @@ def reset_file(file_tracker):
...
     voter_updates = []
     processed_ncids = set()
+
+    def flush():
+        with transaction.atomic():
+            if voter_records:
+                NCVoter.objects.bulk_create(voter_records)
+        with transaction.atomic():
+            for c in change_records:
+                c.voter = c.voter
```
Comment (vkurup) [inline voter/management/commands/voter_process_snapshot.py]: I saw (but didn't quite understand) your conversation about this in HipChat. Could you add a comment explaining why this is needed?

File: voter/management/commands/voter_process_snapshot.py [hunk 7]
```diff
@@ -122,129 +126,139 @@ def reset_file(file_tracker):
...
-                change_records.clear()
-                voter_records.clear()
-                voter_updates.clear()
-                processed_ncids.clear()
-                continue
-            break
+        total_lines += 1
+
+        # If we see a repeat, flushed queued data before continuing
+        # This prevents the same voter from appearing twice in a single bulkd insert
```
Comment (vkurup) [inline voter/management/commands/voter_process_snapshot.py]: good idea!

File: voter/management/commands/voter_process_snapshot.py [hunk 8]
```diff
@@ -122,129 +126,139 @@ def reset_file(file_tracker):
...
             modified_tally += 1
-        change_tracker_values = {
-            'ncid': row['ncid'],
-            'md5_hash': hash_val,
-            'file_tracker': file_tracker,
-            'op_code': change_tracker_op_code,
-            'data': change_tracker_data}
-        if snapshot_dt:
-            change_tracker_values['snapshot_dt'] = snapshot_dt
+            change_tracker_op_code = 'M'
```
Comment (vkurup) [inline voter/management/commands/voter_process_snapshot.py]: let's keep using the ChangeTracker_OP_CODE_MODIFY constant

File: voter/management/commands/voter_process_snapshot.py [hunk 9]
```diff
@@ -122,129 +126,139 @@ def reset_file(file_tracker):
...
+            existing_data = voter_instance.build_current()
+            change_tracker_data = diff_dicts(existing_data, parsed_row)
         else:
-            change_tracker_values['snapshot_dt'] = file_date
-        change_records.append(ChangeTracker(**change_tracker_values))
+            added_tally += 1
+            change_tracker_op_code = 'A'
```
Comment (vkurup) [inline voter/management/commands/voter_process_snapshot.py]: let's use the constant here

File: scripts/mutate_dataset.py (hunk 1 above)
Comment (ironfroggy) [reply scripts/mutate_dataset.py]: I'll add a docstring, it was for taking a snippet of a snapshot and randomly altering some data to use in test fixtures. to simulate the same people appearing with changes over time. I used it two create the 2011 dataset added in this PR.

File: voter/management/commands/voter_process_snapshot.py (hunk 5 above)
Comment (ironfroggy) [reply voter/management/commands/voter_process_snapshot.py]: hmm no i don't want to bring it back. if anything i want to make the imports resumable, which we can't do if the whole thing is one transaction.

Comment (ironfroggy) [reply voter/management/commands/voter_process_snapshot.py]: commented line removed

File: scripts/mutate_dataset.py (hunk 1 above)
Comment (vkurup) [reply scripts/mutate_dataset.py]: :+1:

File: run_tests.sh [hunk 10]
```diff
@@ -6,5 +6,5 @@ flake8 .
 coverage erase
 python manage.py makemigrations --dry-run --check
 # TODO: Uncomment after getting tests running
```
Comment (vkurup) [inline run_tests.sh]: We could just remove the TODO now :)

**Pull request merged**